File: compat/suites/node-js-sdk/src/groups/ecs.ts
/**
 * groups/ecs.ts — ECS compatibility test groups for the Node.js suite.
 *
 * Status: NOT implemented in Overcast. All tests expected to fail with 501.
 * These tests define the coverage target for future ECS implementation.
 *
 * Groups:
 *   ecs-clusters — cluster and task definition lifecycle
 *   ecs-tasks    — task run/describe/list/stop lifecycle
 *   ecs-services — service create/describe/list/update/delete lifecycle
 */

import {
  CreateClusterCommand,
  DeleteClusterCommand,
  DescribeClustersCommand,
  ListClustersCommand,
  RegisterTaskDefinitionCommand,
  DeregisterTaskDefinitionCommand,
  ListTaskDefinitionsCommand,
  RunTaskCommand,
  DescribeTasksCommand,
  ListTasksCommand,
  StopTaskCommand,
  CreateServiceCommand,
  DescribeServicesCommand,
  ListServicesCommand,
  UpdateServiceCommand,
  DeleteServiceCommand,
  NetworkMode,
  LaunchType,
} from "@aws-sdk/client-ecs";
import { makeClients } from "../lib/clients.js";
import type { TestGroup } from "../lib/harness.js";
import * as assert from "node:assert/strict";

export function makeECSGroups(suite: string): TestGroup[] {
  return [
    // ── ecs-clusters ───────────────────────────────────────────────────────
    {
      suite,
      service: "ecs",
      name: "ecs-clusters",
      tests: [
        {
          name: "CreateCluster",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            const clusterName = `compat-${ctx.runId}`;
            const resp = await ecs.send(
              new CreateClusterCommand({ clusterName }),
            );
            assert.ok(resp.cluster?.clusterArn, "CreateCluster: missing clusterArn");
            (ctx as Record<string, unknown>)["_clusterName"] = clusterName;
          },
        },
        {
          name: "DescribeClusters",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            const clusterName = (ctx as Record<string, unknown>)[
              "_clusterName"
            ] as string;
            assert.ok(clusterName, "DescribeClusters: no cluster from CreateCluster");
            const resp = await ecs.send(
              new DescribeClustersCommand({ clusters: [clusterName] }),
            );
            assert.ok(resp.clusters?.length, "DescribeClusters: no clusters returned");
          },
        },
        {
          name: "ListClusters",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            await ecs.send(new ListClustersCommand({}));
          },
        },
        {
          name: "RegisterTaskDefinition",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            const family = `compat-${ctx.runId}`;
            const resp = await ecs.send(
              new RegisterTaskDefinitionCommand({
                family,
                networkMode: NetworkMode.AWSVPC,
                requiresCompatibilities: [LaunchType.FARGATE],
                cpu: "256",
                memory: "512",
                containerDefinitions: [
                  {
                    name: "app",
                    image: "public.ecr.aws/nginx/nginx:latest",
                    essential: true,
                  },
                ],
              }),
            );
            assert.ok(resp.taskDefinition?.taskDefinitionArn, "RegisterTaskDefinition: missing taskDefinitionArn");
            (ctx as Record<string, unknown>)["_taskDefArn"] =
              resp.taskDefinition.taskDefinitionArn;
          },
        },
        {
          name: "ListTaskDefinitions",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            await ecs.send(new ListTaskDefinitionsCommand({}));
          },
        },
        {
          name: "DeregisterTaskDefinition",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            const taskDefArn = (ctx as Record<string, unknown>)[
              "_taskDefArn"
            ] as string;
            if (!taskDefArn) return;
            await ecs.send(
              new DeregisterTaskDefinitionCommand({
                taskDefinition: taskDefArn,
              }),
            );
          },
        },
        {
          name: "DeleteCluster",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            const clusterName = (ctx as Record<string, unknown>)[
              "_clusterName"
            ] as string;
            if (!clusterName) return;
            await ecs.send(new DeleteClusterCommand({ cluster: clusterName }));
          },
        },
      ],
      teardown: async (ctx) => {
        const { ecs } = makeClients(ctx);
        const taskDefArn = (ctx as Record<string, unknown>)[
          "_taskDefArn"
        ] as string;
        if (taskDefArn) {
          try {
            await ecs.send(
              new DeregisterTaskDefinitionCommand({
                taskDefinition: taskDefArn,
              }),
            );
          } catch {}
        }
        const clusterName = (ctx as Record<string, unknown>)[
          "_clusterName"
        ] as string;
        if (clusterName) {
          try {
            await ecs.send(new DeleteClusterCommand({ cluster: clusterName }));
          } catch {}
        }
      },
    },

    // ── ecs-tasks ──────────────────────────────────────────────────────────
    {
      suite,
      service: "ecs",
      name: "ecs-tasks",
      tests: [
        {
          name: "RunTask",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            const clusterName = (ctx as Record<string, unknown>)[
              "_taskCluster"
            ] as string;
            const taskDefArn = (ctx as Record<string, unknown>)[
              "_taskTaskDefArn"
            ] as string;
            assert.ok(clusterName, "RunTask: no cluster from setup");
            assert.ok(taskDefArn, "RunTask: no task definition from setup");
            const resp = await ecs.send(
              new RunTaskCommand({
                cluster: clusterName,
                taskDefinition: taskDefArn,
                launchType: LaunchType.FARGATE,
                networkConfiguration: {
                  awsvpcConfiguration: {
                    subnets: ["subnet-00000000"],
                    assignPublicIp: "DISABLED",
                  },
                },
              }),
            );
            assert.ok(resp.tasks?.length, "RunTask: no tasks returned");
            assert.ok(resp.tasks[0].taskArn, "RunTask: missing taskArn");
            (ctx as Record<string, unknown>)["_taskArn"] =
              resp.tasks[0].taskArn;
          },
        },
        {
          name: "DescribeTasks",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            const clusterName = (ctx as Record<string, unknown>)[
              "_taskCluster"
            ] as string;
            const taskArn = (ctx as Record<string, unknown>)[
              "_taskArn"
            ] as string;
            assert.ok(taskArn, "DescribeTasks: no task from RunTask");
            const resp = await ecs.send(
              new DescribeTasksCommand({
                cluster: clusterName,
                tasks: [taskArn],
              }),
            );
            assert.ok(resp.tasks?.length, "DescribeTasks: no tasks returned");
            assert.ok(
              resp.tasks[0].taskDefinitionArn,
              "DescribeTasks: missing taskDefinitionArn",
            );
          },
        },
        {
          name: "ListTasks",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            const clusterName = (ctx as Record<string, unknown>)[
              "_taskCluster"
            ] as string;
            const taskArn = (ctx as Record<string, unknown>)[
              "_taskArn"
            ] as string;
            const resp = await ecs.send(
              new ListTasksCommand({ cluster: clusterName }),
            );
            assert.ok(resp.taskArns?.length, "ListTasks: empty taskArns");
            assert.ok(
              resp.taskArns.some((a) => a === taskArn),
              "ListTasks: running task not found in list",
            );
          },
        },
        {
          name: "StopTask",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            const clusterName = (ctx as Record<string, unknown>)[
              "_taskCluster"
            ] as string;
            const taskArn = (ctx as Record<string, unknown>)[
              "_taskArn"
            ] as string;
            assert.ok(taskArn, "StopTask: no task from RunTask");
            const resp = await ecs.send(
              new StopTaskCommand({
                cluster: clusterName,
                task: taskArn,
                reason: "compat test cleanup",
              }),
            );
            assert.ok(resp.task, "StopTask: missing task in response");
            assert.strictEqual(
              resp.task.desiredStatus,
              "STOPPED",
              `StopTask: expected desiredStatus STOPPED, got ${resp.task.desiredStatus}`,
            );
          },
        },
      ],
      setup: async (ctx) => {
        const { ecs } = makeClients(ctx);
        const clusterName = `compat-task-${ctx.runId}`;
        await ecs.send(new CreateClusterCommand({ clusterName }));
        (ctx as Record<string, unknown>)["_taskCluster"] = clusterName;
        const family = `compat-task-${ctx.runId}`;
        const resp = await ecs.send(
          new RegisterTaskDefinitionCommand({
            family,
            networkMode: NetworkMode.AWSVPC,
            requiresCompatibilities: [LaunchType.FARGATE],
            cpu: "256",
            memory: "512",
            containerDefinitions: [
              {
                name: "app",
                image: "public.ecr.aws/nginx/nginx:latest",
                essential: true,
              },
            ],
          }),
        );
        (ctx as Record<string, unknown>)["_taskTaskDefArn"] =
          resp.taskDefinition!.taskDefinitionArn!;
      },
      teardown: async (ctx) => {
        const { ecs } = makeClients(ctx);
        const clusterName = (ctx as Record<string, unknown>)[
          "_taskCluster"
        ] as string;
        const taskArn = (ctx as Record<string, unknown>)[
          "_taskArn"
        ] as string;
        if (taskArn && clusterName) {
          try {
            await ecs.send(
              new StopTaskCommand({
                cluster: clusterName,
                task: taskArn,
                reason: "compat teardown",
              }),
            );
          } catch {}
        }
        const taskDefArn = (ctx as Record<string, unknown>)[
          "_taskTaskDefArn"
        ] as string;
        if (taskDefArn) {
          try {
            await ecs.send(
              new DeregisterTaskDefinitionCommand({
                taskDefinition: taskDefArn,
              }),
            );
          } catch {}
        }
        if (clusterName) {
          try {
            await ecs.send(
              new DeleteClusterCommand({ cluster: clusterName }),
            );
          } catch {}
        }
      },
    },

    // ── ecs-services ───────────────────────────────────────────────────────
    {
      suite,
      service: "ecs",
      name: "ecs-services",
      tests: [
        {
          name: "CreateService",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            const clusterName = (ctx as Record<string, unknown>)[
              "_svcCluster"
            ] as string;
            const taskDefArn = (ctx as Record<string, unknown>)[
              "_svcTaskDefArn"
            ] as string;
            assert.ok(clusterName, "CreateService: no cluster from setup");
            assert.ok(taskDefArn, "CreateService: no task definition from setup");
            const serviceName = `compat-svc-${ctx.runId}`;
            const resp = await ecs.send(
              new CreateServiceCommand({
                cluster: clusterName,
                serviceName,
                taskDefinition: taskDefArn,
                desiredCount: 1,
                launchType: LaunchType.FARGATE,
                networkConfiguration: {
                  awsvpcConfiguration: {
                    subnets: ["subnet-00000000"],
                    assignPublicIp: "DISABLED",
                  },
                },
              }),
            );
            assert.ok(
              resp.service?.serviceArn,
              "CreateService: missing serviceArn",
            );
            assert.strictEqual(
              resp.service?.serviceName,
              serviceName,
              `CreateService: expected serviceName ${serviceName}, got ${resp.service?.serviceName}`,
            );
            (ctx as Record<string, unknown>)["_serviceName"] = serviceName;
          },
        },
        {
          name: "DescribeServices",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            const clusterName = (ctx as Record<string, unknown>)[
              "_svcCluster"
            ] as string;
            const serviceName = (ctx as Record<string, unknown>)[
              "_serviceName"
            ] as string;
            assert.ok(serviceName, "DescribeServices: no service from CreateService");
            const resp = await ecs.send(
              new DescribeServicesCommand({
                cluster: clusterName,
                services: [serviceName],
              }),
            );
            assert.ok(
              resp.services?.length,
              "DescribeServices: no services returned",
            );
            assert.strictEqual(
              resp.services[0].serviceName,
              serviceName,
              `DescribeServices: expected serviceName ${serviceName}`,
            );
            assert.ok(
              resp.services[0].desiredCount !== undefined,
              "DescribeServices: missing desiredCount",
            );
            assert.ok(
              resp.services[0].status,
              "DescribeServices: missing status",
            );
          },
        },
        {
          name: "ListServices",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            const clusterName = (ctx as Record<string, unknown>)[
              "_svcCluster"
            ] as string;
            const resp = await ecs.send(
              new ListServicesCommand({ cluster: clusterName }),
            );
            assert.ok(
              resp.serviceArns?.length,
              "ListServices: empty serviceArns",
            );
          },
        },
        {
          name: "UpdateService",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            const clusterName = (ctx as Record<string, unknown>)[
              "_svcCluster"
            ] as string;
            const serviceName = (ctx as Record<string, unknown>)[
              "_serviceName"
            ] as string;
            assert.ok(serviceName, "UpdateService: no service from CreateService");
            const resp = await ecs.send(
              new UpdateServiceCommand({
                cluster: clusterName,
                service: serviceName,
                desiredCount: 2,
              }),
            );
            assert.ok(resp.service, "UpdateService: missing service in response");
            assert.strictEqual(
              resp.service.desiredCount,
              2,
              `UpdateService: expected desiredCount 2, got ${resp.service.desiredCount}`,
            );
          },
        },
        {
          name: "DeleteService",
          fn: async (ctx) => {
            const { ecs } = makeClients(ctx);
            const clusterName = (ctx as Record<string, unknown>)[
              "_svcCluster"
            ] as string;
            const serviceName = (ctx as Record<string, unknown>)[
              "_serviceName"
            ] as string;
            if (!serviceName) return;
            // Must set desiredCount to 0 before deleting
            await ecs.send(
              new UpdateServiceCommand({
                cluster: clusterName,
                service: serviceName,
                desiredCount: 0,
              }),
            );
            await ecs.send(
              new DeleteServiceCommand({
                cluster: clusterName,
                service: serviceName,
              }),
            );
          },
        },
      ],
      setup: async (ctx) => {
        const { ecs } = makeClients(ctx);
        const clusterName = `compat-svc-${ctx.runId}`;
        await ecs.send(new CreateClusterCommand({ clusterName }));
        (ctx as Record<string, unknown>)["_svcCluster"] = clusterName;
        const family = `compat-svc-${ctx.runId}`;
        const resp = await ecs.send(
          new RegisterTaskDefinitionCommand({
            family,
            networkMode: NetworkMode.AWSVPC,
            requiresCompatibilities: [LaunchType.FARGATE],
            cpu: "256",
            memory: "512",
            containerDefinitions: [
              {
                name: "app",
                image: "public.ecr.aws/nginx/nginx:latest",
                essential: true,
              },
            ],
          }),
        );
        (ctx as Record<string, unknown>)["_svcTaskDefArn"] =
          resp.taskDefinition!.taskDefinitionArn!;
      },
      teardown: async (ctx) => {
        const { ecs } = makeClients(ctx);
        const clusterName = (ctx as Record<string, unknown>)[
          "_svcCluster"
        ] as string;
        const serviceName = (ctx as Record<string, unknown>)[
          "_serviceName"
        ] as string;
        if (serviceName && clusterName) {
          try {
            await ecs.send(
              new UpdateServiceCommand({
                cluster: clusterName,
                service: serviceName,
                desiredCount: 0,
              }),
            );
          } catch {}
          try {
            await ecs.send(
              new DeleteServiceCommand({
                cluster: clusterName,
                service: serviceName,
              }),
            );
          } catch {}
        }
        const taskDefArn = (ctx as Record<string, unknown>)[
          "_svcTaskDefArn"
        ] as string;
        if (taskDefArn) {
          try {
            await ecs.send(
              new DeregisterTaskDefinitionCommand({
                taskDefinition: taskDefArn,
              }),
            );
          } catch {}
        }
        if (clusterName) {
          try {
            await ecs.send(
              new DeleteClusterCommand({ cluster: clusterName }),
            );
          } catch {}
        }
      },
    },
  ];
}

File: compat/suites/cli/internal/groups/ecs.go
package groups

import (
	"context"
	"fmt"

	"github.com/Neaox/overcast-compat-cli/internal/awscli"
	"github.com/Neaox/overcast-compat-cli/internal/harness"
)

// ECS returns the ECS service group.
func ECS() ServiceGroup {
	g := &ecsCliGroup{}
	return ServiceGroup{
		Impls: map[string]harness.TestFn{
			"CreateCluster":            g.CreateCluster,
			"DescribeClusters":         g.DescribeClusters,
			"ListClusters":             g.ListClusters,
			"RegisterTaskDefinition":   g.RegisterTaskDefinition,
			"ListTaskDefinitions":      g.ListTaskDefinitions,
			"DeregisterTaskDefinition": g.DeregisterTaskDefinition,
			"DeleteCluster":            g.DeleteCluster,
			"RunTask":                  g.RunTask,
			"DescribeTasks":            g.DescribeTasks,
			"ListTasks":                g.ListTasks,
			"StopTask":                 g.StopTask,
			"CreateService":            g.CreateService,
			"DescribeServices":         g.DescribeServices,
			"ListServices":             g.ListServices,
			"UpdateService":            g.UpdateService,
			"DeleteService":            g.DeleteService,
		},
		Setup: map[string]func(context.Context, *harness.TestContext) error{
			"ecs-clusters": g.setupClusters,
			"ecs-tasks":    g.setupTasks,
			"ecs-services": g.setupServices,
		},
		Teardown: map[string]func(context.Context, *harness.TestContext) error{
			"ecs-clusters": g.teardownClusters,
			"ecs-tasks":    g.teardownTasks,
			"ecs-services": g.teardownServices,
		},
	}
}

type ecsCliGroup struct{}

func (g *ecsCliGroup) setupClusters(_ context.Context, _ *harness.TestContext) error { return nil }
func (g *ecsCliGroup) teardownClusters(_ context.Context, t *harness.TestContext) error {
	if name := t.GetString("ecs_cluster"); name != "" {
		awscli.Run(t.Endpoint, t.Region, "ecs", "delete-cluster", "--cluster", name) //nolint:errcheck
	}
	return nil
}

func (g *ecsCliGroup) CreateCluster(_ context.Context, t *harness.TestContext) error {
	name := fmt.Sprintf("compat-%s", t.RunID)
	out, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "create-cluster",
		"--cluster-name", name,
	)
	if err != nil {
		return err
	}
	cluster, _ := out["cluster"].(map[string]interface{})
	arn, _ := cluster["clusterArn"].(string)
	if arn == "" {
		return fmt.Errorf("CreateCluster: missing clusterArn")
	}
	t.Set("ecs_cluster", name)
	return nil
}

func (g *ecsCliGroup) DescribeClusters(_ context.Context, t *harness.TestContext) error {
	name := t.GetString("ecs_cluster")
	if name == "" {
		return fmt.Errorf("DescribeClusters: no cluster from CreateCluster")
	}
	out, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "describe-clusters",
		"--clusters", name,
	)
	if err != nil {
		return err
	}
	clusters, _ := out["clusters"].([]interface{})
	if len(clusters) == 0 {
		return fmt.Errorf("DescribeClusters: no clusters returned")
	}
	return nil
}

func (g *ecsCliGroup) ListClusters(_ context.Context, t *harness.TestContext) error {
	_, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "list-clusters")
	return err
}

func (g *ecsCliGroup) RegisterTaskDefinition(_ context.Context, t *harness.TestContext) error {
	family := fmt.Sprintf("compat-%s", t.RunID)
	out, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "register-task-definition",
		"--family", family,
		"--network-mode", "bridge",
		"--container-definitions", `[{"name":"app","image":"public.ecr.aws/nginx/nginx:latest","memory":128}]`,
	)
	if err != nil {
		return err
	}
	td, _ := out["taskDefinition"].(map[string]interface{})
	arn, _ := td["taskDefinitionArn"].(string)
	if arn == "" {
		return fmt.Errorf("RegisterTaskDefinition: missing taskDefinitionArn")
	}
	t.Set("task_def_arn", arn)
	return nil
}

func (g *ecsCliGroup) ListTaskDefinitions(_ context.Context, t *harness.TestContext) error {
	_, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "list-task-definitions")
	return err
}

func (g *ecsCliGroup) DeregisterTaskDefinition(_ context.Context, t *harness.TestContext) error {
	arn := t.GetString("task_def_arn")
	if arn == "" {
		return fmt.Errorf("DeregisterTaskDefinition: no task_def_arn from RegisterTaskDefinition")
	}
	return awscli.Run(t.Endpoint, t.Region, "ecs", "deregister-task-definition",
		"--task-definition", arn,
	)
}

func (g *ecsCliGroup) DeleteCluster(_ context.Context, t *harness.TestContext) error {
	name := fmt.Sprintf("compat-del-%s", t.RunID)
	awscli.Run(t.Endpoint, t.Region, "ecs", "create-cluster", "--cluster-name", name) //nolint:errcheck
	return awscli.Run(t.Endpoint, t.Region, "ecs", "delete-cluster",
		"--cluster", name,
	)
}

// ── ecs-tasks ──────────────────────────────────────────────────────────────

func (g *ecsCliGroup) setupTasks(_ context.Context, t *harness.TestContext) error {
	cluster := fmt.Sprintf("compat-task-%s", t.RunID)
	_, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "create-cluster", "--cluster-name", cluster)
	if err != nil {
		return err
	}
	t.Set("task_cluster", cluster)

	family := fmt.Sprintf("compat-task-%s", t.RunID)
	out, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "register-task-definition",
		"--family", family,
		"--network-mode", "awsvpc",
		"--requires-compatibilities", "FARGATE",
		"--cpu", "256",
		"--memory", "512",
		"--container-definitions", `[{"name":"app","image":"public.ecr.aws/nginx/nginx:latest","essential":true}]`,
	)
	if err != nil {
		return err
	}
	td, _ := out["taskDefinition"].(map[string]interface{})
	arn, _ := td["taskDefinitionArn"].(string)
	t.Set("task_taskdef_arn", arn)
	return nil
}

func (g *ecsCliGroup) teardownTasks(_ context.Context, t *harness.TestContext) error {
	if taskArn := t.GetString("task_arn"); taskArn != "" {
		if cluster := t.GetString("task_cluster"); cluster != "" {
			awscli.Run(t.Endpoint, t.Region, "ecs", "stop-task", "--cluster", cluster, "--task", taskArn, "--reason", "teardown") //nolint:errcheck
		}
	}
	if arn := t.GetString("task_taskdef_arn"); arn != "" {
		awscli.Run(t.Endpoint, t.Region, "ecs", "deregister-task-definition", "--task-definition", arn) //nolint:errcheck
	}
	if cluster := t.GetString("task_cluster"); cluster != "" {
		awscli.Run(t.Endpoint, t.Region, "ecs", "delete-cluster", "--cluster", cluster) //nolint:errcheck
	}
	return nil
}

func (g *ecsCliGroup) RunTask(_ context.Context, t *harness.TestContext) error {
	cluster := t.GetString("task_cluster")
	taskDef := t.GetString("task_taskdef_arn")
	if cluster == "" || taskDef == "" {
		return fmt.Errorf("RunTask: missing cluster or task definition from setup")
	}
	out, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "run-task",
		"--cluster", cluster,
		"--task-definition", taskDef,
		"--launch-type", "FARGATE",
		"--network-configuration", `{"awsvpcConfiguration":{"subnets":["subnet-00000000"],"assignPublicIp":"DISABLED"}}`,
	)
	if err != nil {
		return err
	}
	tasks, _ := out["tasks"].([]interface{})
	if len(tasks) == 0 {
		return fmt.Errorf("RunTask: no tasks returned")
	}
	task, _ := tasks[0].(map[string]interface{})
	taskArn, _ := task["taskArn"].(string)
	if taskArn == "" {
		return fmt.Errorf("RunTask: missing taskArn")
	}
	t.Set("task_arn", taskArn)
	return nil
}

func (g *ecsCliGroup) DescribeTasks(_ context.Context, t *harness.TestContext) error {
	cluster := t.GetString("task_cluster")
	taskArn := t.GetString("task_arn")
	if taskArn == "" {
		return fmt.Errorf("DescribeTasks: no task from RunTask")
	}
	out, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "describe-tasks",
		"--cluster", cluster,
		"--tasks", taskArn,
	)
	if err != nil {
		return err
	}
	tasks, _ := out["tasks"].([]interface{})
	if len(tasks) == 0 {
		return fmt.Errorf("DescribeTasks: no tasks returned")
	}
	return nil
}

func (g *ecsCliGroup) ListTasks(_ context.Context, t *harness.TestContext) error {
	cluster := t.GetString("task_cluster")
	out, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "list-tasks",
		"--cluster", cluster,
	)
	if err != nil {
		return err
	}
	arns, _ := out["taskArns"].([]interface{})
	if len(arns) == 0 {
		return fmt.Errorf("ListTasks: empty taskArns")
	}
	return nil
}

func (g *ecsCliGroup) StopTask(_ context.Context, t *harness.TestContext) error {
	cluster := t.GetString("task_cluster")
	taskArn := t.GetString("task_arn")
	if taskArn == "" {
		return fmt.Errorf("StopTask: no task from RunTask")
	}
	out, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "stop-task",
		"--cluster", cluster,
		"--task", taskArn,
		"--reason", "compat test cleanup",
	)
	if err != nil {
		return err
	}
	task, _ := out["task"].(map[string]interface{})
	desired, _ := task["desiredStatus"].(string)
	if desired != "STOPPED" {
		return fmt.Errorf("StopTask: expected desiredStatus STOPPED, got %s", desired)
	}
	return nil
}

// ── ecs-services ───────────────────────────────────────────────────────────

func (g *ecsCliGroup) setupServices(_ context.Context, t *harness.TestContext) error {
	cluster := fmt.Sprintf("compat-svc-%s", t.RunID)
	_, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "create-cluster", "--cluster-name", cluster)
	if err != nil {
		return err
	}
	t.Set("svc_cluster", cluster)

	family := fmt.Sprintf("compat-svc-%s", t.RunID)
	out, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "register-task-definition",
		"--family", family,
		"--network-mode", "awsvpc",
		"--requires-compatibilities", "FARGATE",
		"--cpu", "256",
		"--memory", "512",
		"--container-definitions", `[{"name":"app","image":"public.ecr.aws/nginx/nginx:latest","essential":true}]`,
	)
	if err != nil {
		return err
	}
	td, _ := out["taskDefinition"].(map[string]interface{})
	arn, _ := td["taskDefinitionArn"].(string)
	t.Set("svc_taskdef_arn", arn)
	return nil
}

func (g *ecsCliGroup) teardownServices(_ context.Context, t *harness.TestContext) error {
	svcName := t.GetString("svc_name")
	cluster := t.GetString("svc_cluster")
	if svcName != "" && cluster != "" {
		awscli.Run(t.Endpoint, t.Region, "ecs", "update-service", "--cluster", cluster, "--service", svcName, "--desired-count", "0") //nolint:errcheck
		awscli.Run(t.Endpoint, t.Region, "ecs", "delete-service", "--cluster", cluster, "--service", svcName)                         //nolint:errcheck
	}
	if arn := t.GetString("svc_taskdef_arn"); arn != "" {
		awscli.Run(t.Endpoint, t.Region, "ecs", "deregister-task-definition", "--task-definition", arn) //nolint:errcheck
	}
	if cluster != "" {
		awscli.Run(t.Endpoint, t.Region, "ecs", "delete-cluster", "--cluster", cluster) //nolint:errcheck
	}
	return nil
}

func (g *ecsCliGroup) CreateService(_ context.Context, t *harness.TestContext) error {
	cluster := t.GetString("svc_cluster")
	taskDef := t.GetString("svc_taskdef_arn")
	if cluster == "" || taskDef == "" {
		return fmt.Errorf("CreateService: missing cluster or task definition from setup")
	}
	svcName := fmt.Sprintf("compat-svc-%s", t.RunID)
	out, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "create-service",
		"--cluster", cluster,
		"--service-name", svcName,
		"--task-definition", taskDef,
		"--desired-count", "1",
		"--launch-type", "FARGATE",
		"--network-configuration", `{"awsvpcConfiguration":{"subnets":["subnet-00000000"],"assignPublicIp":"DISABLED"}}`,
	)
	if err != nil {
		return err
	}
	svc, _ := out["service"].(map[string]interface{})
	if svc["serviceArn"] == nil {
		return fmt.Errorf("CreateService: missing serviceArn")
	}
	t.Set("svc_name", svcName)
	return nil
}

func (g *ecsCliGroup) DescribeServices(_ context.Context, t *harness.TestContext) error {
	cluster := t.GetString("svc_cluster")
	svcName := t.GetString("svc_name")
	if svcName == "" {
		return fmt.Errorf("DescribeServices: no service from CreateService")
	}
	out, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "describe-services",
		"--cluster", cluster,
		"--services", svcName,
	)
	if err != nil {
		return err
	}
	services, _ := out["services"].([]interface{})
	if len(services) == 0 {
		return fmt.Errorf("DescribeServices: no services returned")
	}
	return nil
}

func (g *ecsCliGroup) ListServices(_ context.Context, t *harness.TestContext) error {
	cluster := t.GetString("svc_cluster")
	out, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "list-services",
		"--cluster", cluster,
	)
	if err != nil {
		return err
	}
	arns, _ := out["serviceArns"].([]interface{})
	if len(arns) == 0 {
		return fmt.Errorf("ListServices: empty serviceArns")
	}
	return nil
}

func (g *ecsCliGroup) UpdateService(_ context.Context, t *harness.TestContext) error {
	cluster := t.GetString("svc_cluster")
	svcName := t.GetString("svc_name")
	if svcName == "" {
		return fmt.Errorf("UpdateService: no service from CreateService")
	}
	out, err := awscli.RunOutput(t.Endpoint, t.Region, "ecs", "update-service",
		"--cluster", cluster,
		"--service", svcName,
		"--desired-count", "2",
	)
	if err != nil {
		return err
	}
	svc, _ := out["service"].(map[string]interface{})
	if svc == nil {
		return fmt.Errorf("UpdateService: missing service in response")
	}
	return nil
}

func (g *ecsCliGroup) DeleteService(_ context.Context, t *harness.TestContext) error {
	cluster := t.GetString("svc_cluster")
	svcName := t.GetString("svc_name")
	if svcName == "" {
		return nil
	}
	awscli.Run(t.Endpoint, t.Region, "ecs", "update-service", "--cluster", cluster, "--service", svcName, "--desired-count", "0") //nolint:errcheck
	return awscli.Run(t.Endpoint, t.Region, "ecs", "delete-service",
		"--cluster", cluster,
		"--service", svcName,
	)
}
